### PR TITLE
Create Halls C.R.U.D API ENDPOINTS

### LINKED ISSUE #15: [5pts] Create CRUD Api for Halls (admin)
- Get requests to get all Halls 
- Post request to update Halls Infos 
- Delete request to delete a Halls

### PR DESCRIPTION
**Pull request by ichala**
### CRUD API FOR HALLS

1. ### `Get    /admin/halls `
  **_params:_** 
      N/A

2. ### `Patch  /admin/halls `
  **_params:_** 
```JS
{
hall_id: _HALL_ID_
cost:_HALL_COST_
name:_HALL_NAME_
capacity:_HALL_CAPACITY_
description:_HALL_DESCRIPTION_
image:_HALL_IMAGE_LINK_
}
```

3. ### `Delete /admin/halls `
  **_params:_** 
```JS
{
hall_id: _HALL_ID_
}
```

4. ### `POST   /admin/halls `
  **_params:_**
```JS
{
cost:_HALL_COST_
name:_HALL_NAME_
capacity:_HALL_CAPACITY_
description:_HALL_DESCRIPTION_
image:_HALL_IMAGE_LINK_
}
```

Closes https://github.com/ichala/short-let-api/issues/15

### CODE REVIEW

File: app/controllers/halls_controller.rb [hunk 1]
```diff
@@ -1,2 +1,85 @@
...
+  def update_hall
+    if admin?
+      hall = Hall.find(update_params[:hall_id])
+      raise ActiveRecord::RecordNotFound unless hall
+
+      hall.name = update_params[:name]
+      hall.description = update_params[:description]
+      hall.cost = update_params[:cost]
+      hall.capacity = update_params[:capacity]
+      hall.image = update_params[:image]
```
Comment (aimalamiri) [inline app/controllers/halls_controller.rb]: What do you think about something like this?

```ruby
hall = Hall.find(params.permit(:hall_id))

# if now we remove the hall_id from the update_params we can use it normally
hall.update(update_params)

# other code

```

File: config/routes.rb [hunk 2]
```diff
@@ -12,4 +12,9 @@
    #Pending Reservations Routes
     get "/admin/requests/pending", to: "reservations#pending_reservation_admin"
     patch "/admin/requests/pending", to: "reservations#handle_pending_reservation_admin"
+   #Halls C.R.U.D Routes
+    get "/halls", to: "halls#halls_list"
+    patch "/admin/halls", to: "halls#update_hall"
+    delete "/admin/halls", to: "halls#delete_hall"
+    post "/admin/halls", to: "halls#add_hall"
```
Comment (aimalamiri) [inline config/routes.rb:19]: [Optional] `resources` can be used as a shortcut.

**Review by aimalamiri — CHANGES_REQUESTED** [review]
Hi  👋  

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

almost

## Highlights
- Good PR
- Good commit
- All linters passed


## Required Changes ♻️

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the this thread if something is not 100% clear.

File: config/routes.rb (hunk 2 above)
Comment (d2ndjim) [inline config/routes.rb:19]: @ichala , Great Job implementing the code. I think the routes are fine as they are for customization purposes.

**Review by d2ndjim — APPROVED** [review]
@ichala 
approve

* Meeting all requirements

PR Review result:  APPROVED! 👍 👍 👍

File: app/controllers/halls_controller.rb (hunk 1 above)
Comment (EidHachem) [inline app/controllers/halls_controller.rb]: In your `patch` request we need to send all fields to update the hall object, you can use the code below to check if the user wants to update a specific column:

```
      hall.name = update_params[:name] if update_params[:name].present?
      hall.description = update_params[:description] if update_params[:description].present?
      hall.cost = update_params[:cost] if update_params[:cost].present?
      hall.capacity = update_params[:capacity] if update_params[:capacity].present?
      hall.image = update_params[:image] if update_params[:image].present?
```
then all non-present params will keep their previous values.

**Review by EidHachem — CHANGES_REQUESTED** [review]
Hi @ichala  👋  

Good job so far!
There are some suggestions that you might consider, but your code is great




68747470733a2f2f6d65646961322e67697068792e636f6d2f6d656469612f4a713832345239334a734c775a436169534c2f67697068792e676966

## Highlights
- Good PR
- Good commits


## Required Changes ♻️

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the this thread if something is not 100% clear.

File: app/controllers/halls_controller.rb (hunk 1 above)
Comment (ichala) [reply app/controllers/halls_controller.rb]: @EidHachem the params have required anyway in the permit function so in the front-end if the admin doesn't want to change a field he needs to send it as it is

**Review by EidHachem — APPROVED** [review]

**Review by aimalamiri — APPROVED** [review]
### **Hi 👋** 

### **Congratulations! 🎉**

Your project is complete!  There is nothing else to say other than... it's time to merge it 🚀 : 
 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**Pull request merged**